flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java

```diff
@@ -184,4 +189,23 @@ public static void addPrimaryKey(
     });
   }
 
+  private static void validateWatermarkExpression(LogicalType watermarkType) {
+    if (!canBeTimeAttributeType(watermarkType) || getPrecision(watermarkType) > 3) {
+      throw new ValidationException(
+        String.format(
+          "Invalid data type of expression for watermark definition. " +
+            "The field must be of type TIMESTAMP(p)," +
+            " the supported precision 'p' is from 0 to 3, but the watermark expression type is %s",
+          watermarkType));
+    }
+  }
+
+  public static boolean canBeTimeAttributeType(LogicalType logicalType) {
```

Comment:
The flink-table-runtime module of the Flink project has provided [the util method](https://github.com/apache/flink/blob/master/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/TypeCheckUtils.java#L85) to distinguish whether the logical type variable is Timestamp or not.

Reply:
Thanks for your review. I have made changes here.